src/main/scala/org/broadinstitute/hail/stats/FisherExactTest.scala

```diff
@@ -0,0 +1,270 @@
+          if (q <= high) 1d else 0d
+        else if (q >= high) 1d else 0d
+      } else {
+        dnhyper(ncp)
+          .zipWithIndex
+          .filter { case (d, i) => if (upper_tail) support(i) >= q else support(i) <= q }
+          .map { case (d, i) => d }
+          .sum
+      }
+    }
+
+    def mnhyper(ncp: Double): Double = {
+      if (ncp == 0d)
+        low
+      else if (ncp == Double.PositiveInfinity)
+        high
+      else
+        dnhyper(ncp).zipWithIndex.map { case (dnh, i) => dnh * support(i) }.sum
+    }
+
+    def unirootMnHyper(fn: Double => Double, x: Double)(t: Double) = mnhyper(fn(t)) - x
+
+    def unirootPnHyper(fn: Double => Double, x: Int, upper_tail: Boolean, alpha: Double)(t: Double) = pnhyper(x, fn(t), upper_tail) - alpha
+
+    def inverse(x: Double) = divOption(1d, x)
+
+    def mle(x: Double): Option[Double] = {
+      if (x == low)
+        Option(0d)
+      else if (x == high)
+        Option(Double.PositiveInfinity) // Should be infinity
+      else {
+        val mu = mnhyper(1.0)
+        if (mu > x)
+          uniroot(unirootMnHyper(d => d, x) _, 0.0, 1.0)
+        else if (mu < x)
+          uniroot(unirootMnHyper(d => 1 / d, x) _, 2.220446e-16, 1d).flatMap(inverse)
+        else
+          Option(1d)
+      }
+    }
+
+    def ncpLower(x: Int, alpha: Double): Option[Double] = {
+      if (x == low)
+        Option(0d)
+      else {
+        val p = pnhyper(x, 1, upper_tail = true)
+        if (p > alpha)
+          uniroot(unirootPnHyper(d => d, x, true, alpha)_, 0d, 1d)
+        else if (p < alpha)
+          uniroot(unirootPnHyper(d => 1 / d, x, true, alpha) _, 2.220446e-16, 1d).flatMap(inverse)
+        else
+          Option(1d)
+      }
+    }
+
+    def ncpUpper(x: Int, alpha: Double): Option[Double] = {
+      if (x == high) {
+        Option(Double.PositiveInfinity)
+      }
+      else {
+        val p = pnhyper(x, 1)
+        if (p < alpha)
+          uniroot(unirootPnHyper(d => d, x, false, alpha)_, 0d, 1d)
+        else if (p > alpha)
+          uniroot(unirootPnHyper(d => 1 / d, x, false, alpha) _, 2.220446e-16, 1d).flatMap(inverse)
+        else
+          Option(1d)
+      }
+    }
+
+    val pvalue = alternative match {
+      case "less" => pnhyper(numSuccessSample, oddsRatio)
+      case "greater" => pnhyper(numSuccessSample, oddsRatio, upper_tail = true)
+      case "two.sided" =>
+        if (oddsRatio == 0)
+          if (low == numSuccessSample) 1d else 0d
+        else if (oddsRatio == Double.PositiveInfinity)
+          if (high == numSuccessSample) 1d else 0d
+        else {
+          val relErr = 1d + 1e-7
+          val d = dnhyper(oddsRatio)
+          d.filter(_ <= d(numSuccessSample - low) * relErr).sum
+        }
+
+      case _ => throw new IllegalArgumentException("didn't recognize option for alternative. Use one of [less, greater, two.sided]")
+    }
+
+    if (pvalue < 0d || pvalue > 1.000000000002)
+      fatal(s"pvalue not valid probability: $pvalue")
```

Comment:
I agree it should be an assert. I'll remove this unless you think it's a good check to have. Probably not necessary since our test suite is making sure the results are identical to R.